File: adventofcode/2023/day06/day06.zig
const std = @import("std");
const time = std.time;
const print = std.debug.print;
const split = std.mem.splitScalar;
const trim = std.mem.trim;
const ArrayList = std.ArrayList;

fn count_wins(rtime: u64, rdist: u64) u64 {
    const l = for (0..rtime) |t| {
        if (rdist < (t * (rtime - t))) break t;
    } else 0;
    return @intCast(1 + rtime - 2 * l);
}

fn part1(input: [100]u8) !u64 {
    var lit = split(u8, &input, '\n');
    var tit = split(u8, lit.next().?, ':');
    _ = tit.next();
    var tits = split(u8, trim(u8, tit.next().?, " "), ' ');
    var times = ArrayList(u64).init(std.heap.page_allocator);
    while (tits.next()) |ns| {
        const tns = trim(u8, ns, " ");
        if (std.mem.eql(u8, tns, "")) continue;

        const n = std.fmt.parseInt(u64, tns, 10) catch 0;
        try times.append(n);
    }

    var dit = split(u8, lit.next().?, ':');
    _ = dit.next();
    var dits = split(u8, trim(u8, dit.next().?, " "), ' ');
    var dists = ArrayList(u64).init(std.heap.page_allocator);
    while (dits.next()) |ns| {
        const tns = trim(u8, ns, " ");
        if (std.mem.eql(u8, tns, "")) continue;

        const n = std.fmt.parseInt(u64, tns, 10) catch 0;
        try dists.append(n);
    }

    var w: u64 = 1;
    for (0..times.items.len) |i| {
        w *= count_wins(times.items[i], dists.items[i]);
    }
    return w;
}

fn part2(input: [100]u8) !u64 {
    var lit = split(u8, &input, '\n');
    var tit = split(u8, lit.next().?, ':');
    _ = tit.next();
    const tits = trim(u8, tit.next().?, " ");
    var tnumstr = ArrayList(u8).init(std.heap.page_allocator);
    for (tits) |c| {
        if (c == ' ') continue;

        try tnumstr.append(c);
    }
    const tnums = try tnumstr.toOwnedSlice();
    const tnum = std.fmt.parseInt(u64, tnums, 10) catch 0;

    var dit = split(u8, lit.next().?, ':');
    _ = dit.next();
    const dits = trim(u8, dit.next().?, " ");
    var dnumstr = ArrayList(u8).init(std.heap.page_allocator);
    for (dits) |c| {
        if (c == ' ') continue;

        try dnumstr.append(c);
    }
    const dnums = try dnumstr.toOwnedSlice();
    const dnum = std.fmt.parseInt(u64, dnums, 10) catch 0;
    return count_wins(tnum, dnum);
}

pub fn main() !void {
    const s = time.microTimestamp();

    var lines: [100]u8 = undefined;
    _ = try std.io.getStdIn().readAll(&lines);

    const p1 = try part1(lines);
    print("{d}\n", .{p1});

    const p2 = try part2(lines);
    print("{d}\n", .{p2});

    const e = time.microTimestamp();
    print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}

File: adventofcode/2023/day06/day06.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func winCount(time int, dist int) int {
	for t := 0; t <= time; t++ {
		if dist < (t * (time - t)) {
			return 1 + time - (2 * t)
		}
	}
	return 0
}

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func part1(lines []string) int {
	ts := strings.Fields(lines[0])[1:]
	ds := strings.Fields(lines[1])[1:]
	w := 1
	for i := 0; i < len(ts); i++ {
		tn, _ := strconv.Atoi(ts[i])
		dn, _ := strconv.Atoi(ds[i])
		w *= winCount(tn, dn)
	}
	return w
}

func part2(lines []string) int {
	ts := strings.Fields(lines[0])[1:]
	ds := strings.Fields(lines[1])[1:]
	tnums := strings.Join(ts, "")
	dnums := strings.Join(ds, "")
	tnum, _ := strconv.Atoi(tnums)
	dnum, _ := strconv.Atoi(dnums)
	return winCount(tnum, dnum)
}

func main() {
	s := time.Now()
	lines := readInput()
	s1 := time.Now()
	fmt.Println(part1(lines))
	e1 := time.Since(s1)
	s2 := time.Now()
	fmt.Println(part2(lines))
	e2 := time.Since(s2)
	e := time.Since(s)
	fmt.Printf("Time for part1: %s, part2: %s, total: %s\n", e1, e2, e)
}
